app/text_buffer.py

```diff
@@ -141,155 +143,222 @@ def doMoveLines(self, begin, end, to):
-    if self.redoIndex < self.savedAtRedoIndex:
-      self.savedAtRedoIndex = -1
-    self.redoChain = self.redoChain[:self.redoIndex]
+    # Trim only when there is a non-trivial action.
+    if change[0] == 'm' and self.redoIndex != len(self.redoChain):
+      newTrivialChange = True
+    else:
+      # Trim and combine main redoChain with tempChange
+      # if there is a non-trivial action
+      if self.redoIndex < self.savedAtRedoIndex:
+        self.savedAtRedoIndex = -1
+      self.redoChain = self.redoChain[:self.redoIndex]
+      if self.tempChange:
+        if self.redoChain[-1][0] == 'm':
+          combinedChange = (change[0], addVectors(self.tempChange[1], self.redoChain[-1][1]))
+          if noOpInstruction(combinedChange):
+            self.redoChain.pop()
+            self.redoIndex -= 1
+          else:
+            self.redoChain[self.redoIndex - 1] = combinedChange
+        else:
+          self.redoChain.append(self.tempChange)
+          self.redoIndex += 1
+        self.tempChange = None
     if 1: # optimizer
       if len(self.redoChain) and self.savedAtRedoIndex != self.redoIndex:
         if (self.redoChain[-1][0] == change[0] and
             change[0] in ('d', 'i')):
           change = (change[0], self.redoChain[-1][1] + change[1])
           self.undoOne()
           self.redoChain.pop()
-        elif change[0] == 'm':
-          if self.redoChain[-1][0] == 'm':
-            change = (change[0], addVectors(self.redoChain[-1][1], change[1]))
-            self.undoOne()
-            self.redoChain.pop()
         elif self.redoChain[-1][0] == change[0] and change[0] == 'n':
           change = (change[0], addVectors(self.redoChain[-1][1], change[1]))
           self.undoOne()
           self.redoChain.pop()
-    if 1:
-      # Eliminate no-op entries
-      noOpInstructions = set([
-        ('m', (0,0,0,0,0)),
-      ])
-      assert ('m', (0,0,0,0,0)) in noOpInstructions
-      if change in noOpInstructions:
+    if newTrivialChange:
+      if self.tempChange:
+        #Combine new change with the existing tempChange
+        change = (change[0], addVectors(self.tempChange[1], change[1]))
+        self.undoOne()
+      if noOpInstruction(change):
+        self.tempChange = None
         return
-      #app.log.info('opti', change)
-    self.redoChain.append(change)
+      self.redoDirty = True
+      self.tempChange = change
+    else:
+      if len(self.redoChain) and change[0] == 'm':
+        if self.redoChain[-1][0] == 'm':
+          change = (change[0], addVectors(self.redoChain[-1][1], change[1]))
+          self.undoOne()
+          self.redoChain.pop()
+        if noOpInstruction(change):
+          return
+      self.redoChain.append(change)
     if self.debugRedo:
       app.log.info('--- redoIndex', self.redoIndex)
       for i,c in enumerate(self.redoChain):
         app.log.info('%2d:'%i, repr(c))
+      app.log.info('tempChange', repr(self.tempChange))
+
+  def undoDirty(self):
```

Comment:
Yes, I think I can do that. Will have to remember to manually change tempChange to None after this call though. I'll fix this soon.